File: python/sol/PmlHy.py
# -*- coding: utf-8 -*-
"""
PmlHy.py
"""

from numba import jit, prange

@jit(cache=True, nogil=True, parallel=True, nopython=True)
def calHy(
    Nz, Nx, Hy, Ez, Ex, Hyz, Hyx, RZc, RXc,
    iPmlHy, gPmlZc, gPmlXc, rPmlH, rPml, pml_l,
    Ni, Nj, Nk, N0):

    for n in prange(iPmlHy.shape[0]):
        i, j, k, m = iPmlHy[n]

        n0   = Ni * (i    ) + Nj * (j    ) + Nk * (k    ) + N0
        nk1  = Ni * (i    ) + Nj * (j    ) + Nk * (k + 1) + N0
        ni1  = Ni * (i + 1) + Nj * (j    ) + Nk * (k    ) + N0

        kc = min(max(k, 0), Nz - 1)
        dex = Ex[nk1] - Ex[n0]
        rz = RZc[kc] * rPmlH[m]
        Hyz[n] = (Hyz[n] - (rz * dex)) / (1 + (gPmlZc[k + pml_l] * rPml[m]))

        ic = min(max(i, 0), Nx - 1)
        dez = Ez[ni1] - Ez[n0]
        rx = RXc[ic] * rPmlH[m]
        Hyx[n] = (Hyx[n] + (rx * dez)) / (1 + (gPmlXc[i + pml_l] * rPml[m]))

        Hy[n0] = Hyz[n] + Hyx[n]

"""
# 配列演算(数十倍遅い)
def calHy( \
    Nz, Nx, Hy, Ez, Ex, Hyz, Hyx, RZc, RXc,
    iPmlHy, gPmlZc, gPmlXc, rPmlH, rPml, pml_l,
    Ni, Nj, Nk, N0):
    # [:]が必要

    i = iPmlHy[:, 0]
    j = iPmlHy[:, 1]
    k = iPmlHy[:, 2]
    m = iPmlHy[:, 3]

    n0   = Ni * (i    ) + Nj * (j    ) + Nk * (k    ) + N0
    nk1  = Ni * (i    ) + Nj * (j    ) + Nk * (k + 1) + N0
    ni1  = Ni * (i + 1) + Nj * (j    ) + Nk * (k    ) + N0

    kc = np.minimum(np.maximum(k, 0), Nz - 1)
    dex = Ex[nk1] - Ex[n0]
    rz = RZc[kc] * rPmlH[m]
    Hyz[:] = (Hyz[:] - (rz * dex)) / (1 + (gPmlZc[k + pml_l] * rPml[m]))

    ic = np.minimum(np.maximum(i, 0), Nx - 1)
    dez = Ez[ni1] - Ez[n0]
    rx = RXc[ic] * rPmlH[m]
    Hyx[:] = (Hyx[:] + (rx * dez)) / (1 + (gPmlXc[i + pml_l] * rPml[m]))

    Hy[n0] = Hyz[:] + Hyx[:]
"""

File: sol/pmlHy.c
/*
pmlHy.c (OpenMP)

PML for Hy
*/

#include "ofd.h"

void pmlHy(void)
{
	const int lz = cPML.l;
	const int lx = cPML.l;

	int64_t n;
#ifdef _OPENMP
#pragma omp parallel for
#endif
#ifdef __CLANG_FUJITSU
#pragma clang loop vectorize(assume_safety)
#endif
	for (n = 0; n < numPmlHy; n++) {
		const int  i = fPmlHy[n].i;
		const int  j = fPmlHy[n].j;
		const int  k = fPmlHy[n].k;
		const id_t m = fPmlHy[n].m;

		const real_t dex = EX(i, j, k + 1) - EX(i, j, k);
		const real_t rz = RZc[MIN(MAX(k, 0), Nz - 1)] * rPmlH[m];
		Hyz[n] = (Hyz[n] - (rz * dex)) / (1 + (gPmlZc[k + lz] * rPml[m]));

		const real_t dez = EZ(i + 1, j, k) - EZ(i, j, k);
		const real_t rx = RXc[MIN(MAX(i, 0), Nx - 1)] * rPmlH[m];
		Hyx[n] = (Hyx[n] + (rx * dez)) / (1 + (gPmlXc[i + lx] * rPml[m]));

		HY(i, j, k) = Hyz[n] + Hyx[n];
	}
}
